# Imputation method of spectra

## Preview

In [158]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages
import random
csv_file_path = '02-BaseNew.csv'
df = pd.read_csv(csv_file_path)
df_spectra = df.iloc[:, 0:53]
albedo_column = df.iloc[:, 53]
df

0.45     0.475       0.5     0.525      0.55     0.575       0.6  \
0          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
3          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
4          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
...        ...       ...       ...       ...       ...       ...       ...   
2935       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2936       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2937       NaN       NaN       NaN -0.760226 -0.727909 -0.699084 -0.676065   
2938 -0.424504 -0.432160 -0.434251 -0.432117 -0.432182 -0.426685 -0.424460   
2939 -0.653730 -0.619093 -0.590577 -0.566641 -0.545942 -0.527398 -0.508110   

         0.625      0.65     0.675  ...      2.25       2.3      2.35  \
0          NaN       NaN       NaN  ...  0.178308  0.179383  0.177941   
1          NaN       NaN       NaN  ...  0.231438  0.232110  0.218402   
2          NaN       NaN       NaN  ...  0.220376  0.226100  0.222016   
3          NaN       NaN       NaN  ...  0.154970  0.151323  0.144752   
4          NaN       NaN       NaN  ...  0.210976  0.201246       NaN   
...        ...       ...       ...  ...       ...       ...       ...   
2935       NaN       NaN       NaN  ...  0.320975  0.337816  0.349726   
2936       NaN       NaN       NaN  ...  0.279271  0.297687  0.316016   
2937 -0.651195 -0.622867 -0.595565  ...  0.217393  0.225075  0.236595   
2938 -0.413358 -0.403189 -0.394357  ...  0.240991  0.251120  0.261629   
2939 -0.489623 -0.471062 -0.451295  ...  0.289864  0.302467  0.315095   

           2.4      2.45        pV        name  counts  class_bdm  \
0     0.173358  0.168305       NaN     1988 TA       1          A   
1          NaN       NaN -0.391474    1990 WO3       1          A   
2     0.222024  0.240588       NaN    1993 FO6       1          A   
3     0.138640       NaN -0.655608  1998 FE118       1         Sa   
4          NaN       NaN -0.422508    1998 YG8       1          A   
...        ...       ...       ...         ...     ...        ...   
2935  0.362664       NaN -0.657577     Skorina       3          D   
2936  0.332846  0.346252 -0.958607     Skorina       1          D   
2937  0.251763  0.275564 -1.173925     Thestor       1          D   
2938  0.270172  0.274471 -1.356547       Thule       1          D   
2939  0.328036       NaN -1.259637      Tornio       1          D   

      class_asteroid_sf  
0                     A  
1                     A  
2                     A  
3                     A  
4                     A  
...                 ...  
2935                  Z  
2936                  Z  
2937                  Z  
2938                  Z  
2939                  Z  

[2940 rows x 58 columns]

- Tranform the spectra in reverse of their presented natural logarithm-transformed state.
- Similarly, transform the albedo values in reverse of the transformed log base 10.
- (Both transformations were required by Mahlke, 2022.

In [160]:
# Reverse natural logarithm transformation for spectra
df.iloc[:, 0:53] = np.exp(df_spectra)
# Reverse log base 10 transformation for albedo
df.iloc[:, 53] = 10**albedo_column
df

0.45     0.475       0.5     0.525      0.55     0.575       0.6  \
0          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
3          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
4          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
...        ...       ...       ...       ...       ...       ...       ...   
2935       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2936       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2937       NaN       NaN       NaN  0.467561  0.482917  0.497040  0.508614   
2938  0.654094  0.649106  0.647750  0.649133  0.649091  0.652669  0.654123   
2939  0.520102  0.538432  0.554007  0.567428  0.579296  0.590139  0.601631   

         0.625      0.65     0.675  ...      2.25       2.3      2.35  \
0          NaN       NaN       NaN  ...  1.195193  1.196479  1.194755   
1          NaN       NaN       NaN  ...  1.260412  1.261258  1.244087   
2          NaN       NaN       NaN  ...  1.246545  1.253701  1.248591   
3          NaN       NaN       NaN  ...  1.167623  1.163372  1.155753   
4          NaN       NaN       NaN  ...  1.234882  1.222925       NaN   
...        ...       ...       ...  ...       ...       ...       ...   
2935       NaN       NaN       NaN  ...  1.378472  1.401883  1.418678   
2936       NaN       NaN       NaN  ...  1.322165  1.346740  1.371653   
2937  0.521423  0.536404  0.551251  ...  1.242833  1.252417  1.266928   
2938  0.661425  0.668186  0.674114  ...  1.272509  1.285464  1.299044   
2939  0.612858  0.624339  0.636803  ...  1.336246  1.353193  1.370389   

           2.4      2.45     pV        name  counts  class_bdm  \
0     1.189291  1.183298    NaN     1988 TA       1          A   
1          NaN       NaN  0.406    1990 WO3       1          A   
2     1.248602  1.271996    NaN    1993 FO6       1          A   
3     1.148710       NaN  0.221  1998 FE118       1         Sa   
4          NaN       NaN  0.378    1998 YG8       1          A   
...        ...       ...    ...         ...     ...        ...   
2935  1.437153       NaN  0.220     Skorina       3          D   
2936  1.394933  1.413759  0.110     Skorina       1          D   
2937  1.286292  1.317274  0.067     Thestor       1          D   
2938  1.310190  1.315834  0.044       Thule       1          D   
2939  1.388239       NaN  0.055      Tornio       1          D   

      class_asteroid_sf  
0                     A  
1                     A  
2                     A  
3                     A  
4                     A  
...                 ...  
2935                  Z  
2936                  Z  
2937                  Z  
2938                  Z  
2939                  Z  

[2940 rows x 58 columns]

## Final method - Spectra

In [9]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages
import random

# Set global font to DejaVu Serif and increase font sizes
plt.rcParams['font.family'] = 'DejaVu Serif'
plt.rcParams['font.size'] = 20

# ---------------------------
# Load Data
# ---------------------------

# Assume spectral data are in the first 53 columns.
df_spectra = df.iloc[:, 0:53].copy()
total_cols = df_spectra.shape[1]  # typically 53
wavelengths = np.array([float(col) for col in df_spectra.columns])

# ---------------------------
# Load class labels from column "class_asteroid_sf".
# ---------------------------
if 'class_asteroid_sf' in df.columns:
    classes = df['class_asteroid_sf']
else:
    raise ValueError("No 'class_asteroid_sf' column found in the CSV data.")

# ---------------------------
# Imputation Parameters
# ---------------------------
overlap_points = 21   # number of points used for overlapping region
slope_weight = 1.0    # slope weight = 1 for all imputation

# ---------------------------
# Function to compute error metric for a candidate.
# Uses:
#   - MSE computed with average shift over the overlapping region.
#   - Slope difference computed after aligning using the "i point":
#       * For left-incomplete: use the first overlapping point.
#       * For right-incomplete: use the last overlapping point.
# ---------------------------
def compute_similarity_aligned(target, candidate, indices, slope_weight, side="left"):
    shift_avg = np.mean(target[indices] - candidate[indices])
    candidate_aligned_avg = candidate + shift_avg
    mse = np.mean((target[indices] - candidate_aligned_avg[indices])**2)
    
    if side == "left":
        shift_point = target[indices[0]] - candidate[indices[0]]
    else:
        shift_point = target[indices[-1]] - candidate[indices[-1]]
    candidate_aligned = candidate + shift_point
    if len(indices) > 1:
        target_slopes = np.diff(target[indices])
        candidate_slopes = np.diff(candidate_aligned[indices])
        slope_diff = np.mean(np.abs(target_slopes - candidate_slopes))
    else:
        slope_diff = 0.0
    total_error = mse + slope_weight * slope_diff
    return total_error

# ---------------------------
# Process all spectra and perform imputation for incomplete ones.
# We'll store the final imputed spectra in final_imputed_list.
# ---------------------------
final_imputed_list = [None] * len(df_spectra)

# Export a single PDF with one page per imputed (incomplete) spectrum.
pdf = PdfPages("all_imputed.pdf")

for i in range(len(df_spectra)):
    sample_orig = df_spectra.iloc[i].values.astype(float)
    # Use original for candidate selection.
    sample = sample_orig.copy()
    # If spectrum is complete, leave it unchanged. Here the iteration starts.
    if not (np.isnan(sample[0]) or np.isnan(sample[-1])):
        final_imputed_list[i] = sample
        continue

    final_imputed = sample_orig.copy()  # working copy for final imputation
    processed_left = False
    processed_right = False

    # ---- Left-side imputation ----
    if np.isnan(sample[0]):
        processed_left = True
        first_obs = np.where(~np.isnan(sample))[0][0]
        last_ob = np.where(~np.isnan(sample))[0][-1]
        left_missing_indices = np.arange(0, first_obs)
        left_overlap_indices = np.arange(first_obs, last_ob + 1)
         
        # Find candidate spectra with complete data in left missing & overlapping regions,
        # and from the same class as the target.
        candidates_left = []
        for j in range(len(df_spectra)):
            if j == i:
                continue
            if classes[j] != classes[i]:
                continue
            cand = df_spectra.iloc[j].values.astype(float)
            if np.all(~np.isnan(cand[left_overlap_indices])) and np.all(~np.isnan(cand[left_missing_indices])):
                candidates_left.append((j, cand))
        
        # --- For targets of class "R", use all available candidates even if fewer than 10.
        if classes[i] == "R":
            if len(candidates_left) >= 1:
                candidates_with_error_left = []
                for cand_index, cand in candidates_left:
                    err = compute_similarity_aligned(sample_orig, cand, left_overlap_indices, slope_weight, side="left")
                    candidates_with_error_left.append((cand_index, cand, err))
                candidates_with_error_left.sort(key=lambda x: x[2])
                best_candidates_left = candidates_with_error_left  # use all available candidates
                imputed_left = []
                errors_left = []
                for cand_index, cand, err in best_candidates_left:
                    errors_left.append(err)
                    shift_point = sample_orig[left_overlap_indices[0]] - cand[left_overlap_indices[0]]
                    cand_aligned = cand + shift_point
                    imputed_left.append(cand_aligned[left_missing_indices])
                imputed_left = np.array(imputed_left)
                errors_left = np.array(errors_left)
                weights_left = 1.0 / (errors_left + 1e-6)
                weighted_imputed_left = np.average(imputed_left, axis=0, weights=weights_left)
                
                # Smoothing: blend the first four missing points with extrapolated trend.
                if len(left_overlap_indices) >= 2:
                    x0 = wavelengths[left_overlap_indices[0]]
                    x1 = wavelengths[left_overlap_indices[1]]
                    y0 = sample_orig[left_overlap_indices[0]]
                    y1 = sample_orig[left_overlap_indices[1]]
                    slope_left = (y1 - y0) / (x1 - x0)
                else:
                    slope_left = 0
                smoothed_left = weighted_imputed_left.copy()
                for idx_missing, m in enumerate(left_missing_indices):
                    d = first_obs - m  # distance from the boundary
                    if d == 1:
                        w = 0.5
                    elif d == 2:
                        w = 0.35
                    elif d == 3:
                        w = 0.25
                    elif d == 4:
                        w = 0.15
                    elif d == 5:
                        w = 0.1
                    elif d == 6:
                        w = 0.05
                    elif d == 7:
                        w = 0.0
                    else:
                        w = 0.0
                    if w > 0:
                        extrapolated_val = sample_orig[left_overlap_indices[0]] - slope_left * (x0 - wavelengths[m])
                        smoothed_left[idx_missing] = w * extrapolated_val + (1 - w) * weighted_imputed_left[idx_missing]
                weighted_imputed_left = smoothed_left
                
                final_imputed[left_missing_indices] = weighted_imputed_left
                
                # Plot left-side imputation.
                fig, ax = plt.subplots(figsize=(10,6))
                for cand_index, cand, err in best_candidates_left:
                    shift_point = sample_orig[left_overlap_indices[0]] - cand[left_overlap_indices[0]]
                    cand_aligned = cand + shift_point
                    ax.plot(wavelengths, cand_aligned, color='lightgray', linewidth=1, zorder=1)
                ax.plot(wavelengths, sample_orig, 'ko-', label="Target (Observed)", zorder=3)
                ax.plot(wavelengths, final_imputed, 'b--', label="Final Imputed Spectrum", zorder=4)
                ax.scatter(wavelengths[left_missing_indices], final_imputed[left_missing_indices], color='red', 
                           label="Imputed Points", zorder=5)
                ax.axvspan(wavelengths[left_missing_indices[0]], wavelengths[left_missing_indices[-1]], 
                           color='red', alpha=0.2, label="Missing Region", zorder=2)
                ax.set_xlabel("Wavelength (µm)")
                ax.set_ylabel("Reflectance")
                ax.legend()
                ax.text(0.5, -0.2, f"Target {i}: Left Imputation", transform=ax.transAxes, 
                        ha='center', va='center', fontsize=20)
                pdf.savefig(fig, bbox_inches='tight')
                plt.close()
        else:
            if len(candidates_left) < 10:
                pass  # Skip left-imputation if not enough candidates.
            else:
                candidates_with_error_left = []
                for cand_index, cand in candidates_left:
                    err = compute_similarity_aligned(sample_orig, cand, left_overlap_indices, slope_weight, side="left")
                    candidates_with_error_left.append((cand_index, cand, err))
                candidates_with_error_left.sort(key=lambda x: x[2])
                best_candidates_left = candidates_with_error_left[:10]
                
                imputed_left = []
                errors_left = []
                for cand_index, cand, err in best_candidates_left:
                    errors_left.append(err)
                    shift_point = sample_orig[left_overlap_indices[0]] - cand[left_overlap_indices[0]]
                    cand_aligned = cand + shift_point
                    imputed_left.append(cand_aligned[left_missing_indices])
                imputed_left = np.array(imputed_left)
                errors_left = np.array(errors_left)
                weights_left = 1.0 / (errors_left + 1e-6)
                weighted_imputed_left = np.average(imputed_left, axis=0, weights=weights_left)
                
                if len(left_overlap_indices) >= 2:
                    x0 = wavelengths[left_overlap_indices[0]]
                    x1 = wavelengths[left_overlap_indices[1]]
                    y0 = sample_orig[left_overlap_indices[0]]
                    y1 = sample_orig[left_overlap_indices[1]]
                    slope_left = (y1 - y0) / (x1 - x0)
                else:
                    slope_left = 0
                smoothed_left = weighted_imputed_left.copy()
                for idx_missing, m in enumerate(left_missing_indices):
                    d = first_obs - m
                    if d == 1:
                        w = 0.5
                    elif d == 2:
                        w = 0.35
                    elif d == 3:
                        w = 0.25
                    elif d == 4:
                        w = 0.15
                    elif d == 5:
                        w = 0.1
                    elif d == 6:
                        w = 0.05
                    elif d == 7:
                        w = 0.0
                    else:
                        w = 0.0
                    if w > 0:
                        extrapolated_val = sample_orig[left_overlap_indices[0]] - slope_left * (x0 - wavelengths[m])
                        smoothed_left[idx_missing] = w * extrapolated_val + (1 - w) * weighted_imputed_left[idx_missing]
                weighted_imputed_left = smoothed_left
                
                final_imputed[left_missing_indices] = weighted_imputed_left
                
                fig, ax = plt.subplots(figsize=(10,6))
                for cand_index, cand, err in best_candidates_left:
                    shift_point = sample_orig[left_overlap_indices[0]] - cand[left_overlap_indices[0]]
                    cand_aligned = cand + shift_point
                    ax.plot(wavelengths, cand_aligned, color='lightgray', linewidth=1, zorder=1)
                ax.plot(wavelengths, sample_orig, 'ko-', label="Target (Observed)", zorder=3)
                ax.plot(wavelengths, final_imputed, 'b--', label="Final Imputed Spectrum", zorder=4)
                ax.scatter(wavelengths[left_missing_indices], final_imputed[left_missing_indices], color='red', 
                           label="Imputed Points", zorder=5)
                ax.axvspan(wavelengths[left_missing_indices[0]], wavelengths[left_missing_indices[-1]], 
                           color='red', alpha=0.2, label="Missing Region", zorder=2)
                ax.set_xlabel("Wavelength (µm)")
                ax.set_ylabel("Reflectance")
                ax.legend()
                ax.text(0.5, -0.2, f"Target {i}: Left Imputation", transform=ax.transAxes, 
                        ha='center', va='center', fontsize=20)
                pdf.savefig(fig, bbox_inches='tight')
                plt.close()
                
    # ---- Right-side imputation ----
    if np.isnan(sample[-1]):
        processed_right = True
        last_obs = np.where(~np.isnan(sample))[0][-1]
        first_ob = np.where(~np.isnan(sample))[0][0]
        right_missing_indices = np.arange(last_obs + 1, total_cols)
        right_overlap_indices = np.arange(first_ob, last_obs + 1)
        
        candidates_right = []
        for j in range(len(df_spectra)):
            if j == i:
                continue
            if classes[j] != classes[i]:
                continue
            cand = df_spectra.iloc[j].values.astype(float)
            if np.all(~np.isnan(cand[right_overlap_indices])) and np.all(~np.isnan(cand[right_missing_indices])):
                candidates_right.append((j, cand))
                
        if classes[i] == "R":
            if len(candidates_right) >= 1:
                candidates_with_error_right = []
                for cand_index, cand in candidates_right:
                    err = compute_similarity_aligned(sample_orig, cand, right_overlap_indices, slope_weight, side="right")
                    candidates_with_error_right.append((cand_index, cand, err))
                candidates_with_error_right.sort(key=lambda x: x[2])
                best_candidates_right = candidates_with_error_right  # use all available candidates
                imputed_right = []
                errors_right = []
                for cand_index, cand, err in best_candidates_right:
                    errors_right.append(err)
                    shift_point = sample_orig[right_overlap_indices[-1]] - cand[right_overlap_indices[-1]]
                    cand_aligned = cand + shift_point
                    imputed_right.append(cand_aligned[right_missing_indices])
                imputed_right = np.array(imputed_right)
                errors_right = np.array(errors_right)
                weights_right = 1.0 / (errors_right + 1e-6)
                weighted_imputed_right = np.average(imputed_right, axis=0, weights=weights_right)
                
                if len(right_overlap_indices) >= 2:
                    x0 = wavelengths[right_overlap_indices[-2]]
                    x1 = wavelengths[right_overlap_indices[-1]]
                    y0 = sample_orig[right_overlap_indices[-2]]
                    y1 = sample_orig[right_overlap_indices[-1]]
                    slope_right = (y1 - y0) / (x1 - x0)
                else:
                    slope_right = 0
                smoothed_right = weighted_imputed_right.copy()
                for idx_missing, m in enumerate(right_missing_indices):
                    d = m - last_obs  # distance from boundary
                    if d == 1:
                        w = 0.5
                    elif d == 2:
                        w = 0.35
                    elif d == 3:
                        w = 0.25
                    elif d == 4:
                        w = 0.15
                    elif d == 5:
                        w = 0.1
                    elif d == 6:
                        w = 0.05
                    elif d == 7:
                        w = 0.0
                    else:
                        w = 0.0
                    if w > 0:
                        extrapolated_val = sample_orig[right_overlap_indices[-1]] + slope_right * (wavelengths[m] - wavelengths[right_overlap_indices[-1]])
                        smoothed_right[idx_missing] = w * extrapolated_val + (1 - w) * weighted_imputed_right[idx_missing]
                weighted_imputed_right = smoothed_right
                final_imputed[right_missing_indices] = weighted_imputed_right
                
                fig, ax = plt.subplots(figsize=(10,6))
                for cand_index, cand, err in best_candidates_right:
                    shift_point = sample_orig[right_overlap_indices[-1]] - cand[right_overlap_indices[-1]]
                    cand_aligned = cand + shift_point
                    ax.plot(wavelengths, cand_aligned, color='lightgray', linewidth=1, zorder=1)
                ax.plot(wavelengths, sample_orig, 'ko-', label="Target (Observed)", zorder=3)
                ax.plot(wavelengths, final_imputed, 'b--', label="Final Imputed Spectrum", zorder=4)
                ax.scatter(wavelengths[right_missing_indices], final_imputed[right_missing_indices], color='red', 
                           label="Imputed Points", zorder=5)
                ax.axvspan(wavelengths[right_missing_indices[0]], wavelengths[right_missing_indices[-1]], 
                           color='red', alpha=0.2, label="Missing Region", zorder=2)
                ax.set_xlabel("Wavelength (µm)")
                ax.set_ylabel("Reflectance")
                ax.legend()
                ax.text(0.5, -0.2, f"Target {i}: Right Imputation", transform=ax.transAxes, 
                        ha='center', va='center', fontsize=20)
                pdf.savefig(fig, bbox_inches='tight')
                plt.close()
        else:
            if len(candidates_right) < 10:
                pass
            else:
                candidates_with_error_right = []
                for cand_index, cand in candidates_right:
                    err = compute_similarity_aligned(sample_orig, cand, right_overlap_indices, slope_weight, side="right")
                    candidates_with_error_right.append((cand_index, cand, err))
                candidates_with_error_right.sort(key=lambda x: x[2])
                best_candidates_right = candidates_with_error_right[:10]
                
                imputed_right = []
                errors_right = []
                for cand_index, cand, err in best_candidates_right:
                    errors_right.append(err)
                    shift_point = sample_orig[right_overlap_indices[-1]] - cand[right_overlap_indices[-1]]
                    cand_aligned = cand + shift_point
                    imputed_right.append(cand_aligned[right_missing_indices])
                imputed_right = np.array(imputed_right)
                errors_right = np.array(errors_right)
                weights_right = 1.0 / (errors_right + 1e-6)
                weighted_imputed_right = np.average(imputed_right, axis=0, weights=weights_right)
                
                if len(right_overlap_indices) >= 2:
                    x0 = wavelengths[right_overlap_indices[-2]]
                    x1 = wavelengths[right_overlap_indices[-1]]
                    y0 = sample_orig[right_overlap_indices[-2]]
                    y1 = sample_orig[right_overlap_indices[-1]]
                    slope_right = (y1 - y0) / (x1 - x0)
                else:
                    slope_right = 0
                smoothed_right = weighted_imputed_right.copy()
                for idx_missing, m in enumerate(right_missing_indices):
                    d = m - last_obs
                    if d == 1:
                        w = 0.5
                    elif d == 2:
                        w = 0.35
                    elif d == 3:
                        w = 0.25
                    elif d == 4:
                        w = 0.15
                    elif d == 5:
                        w = 0.1
                    elif d == 6:
                        w = 0.05
                    elif d == 7:
                        w = 0.0
                    else:
                        w = 0.0
                    if w > 0:
                        extrapolated_val = sample_orig[right_overlap_indices[-1]] + slope_right * (wavelengths[m] - wavelengths[right_overlap_indices[-1]])
                        smoothed_right[idx_missing] = w * extrapolated_val + (1 - w) * weighted_imputed_right[idx_missing]
                weighted_imputed_right = smoothed_right
                final_imputed[right_missing_indices] = weighted_imputed_right
                
                fig, ax = plt.subplots(figsize=(10,6))
                for cand_index, cand, err in best_candidates_right:
                    shift_point = sample_orig[right_overlap_indices[-1]] - cand[right_overlap_indices[-1]]
                    cand_aligned = cand + shift_point
                    ax.plot(wavelengths, cand_aligned, color='lightgray', linewidth=1, zorder=1)
                ax.plot(wavelengths, sample_orig, 'ko-', label="Target (Observed)", zorder=3)
                ax.plot(wavelengths, final_imputed, 'b--', label="Final Imputed Spectrum", zorder=4)
                ax.scatter(wavelengths[right_missing_indices], final_imputed[right_missing_indices], color='red', 
                           label="Imputed Points", zorder=5)
                ax.axvspan(wavelengths[right_missing_indices[0]], wavelengths[right_missing_indices[-1]], 
                           color='red', alpha=0.2, label="Missing Region", zorder=2)
                ax.set_xlabel("Wavelength (µm)")
                ax.set_ylabel("Reflectance")
                ax.legend()
                ax.text(0.5, -0.2, f"Target {i}: Right Imputation", transform=ax.transAxes, 
                        ha='center', va='center', fontsize=20)
                pdf.savefig(fig, bbox_inches='tight')
                plt.close()
    
    final_imputed_list[i] = final_imputed

pdf.close()

# ---------------------------
# Export final imputed spectra to CSV
# ---------------------------
df_imputed = pd.DataFrame(final_imputed_list, columns=df_spectra.columns, index=df_spectra.index)
df_imputed.to_csv("03-Base-imputedNEW.csv", index=False)

print("PDF of all imputed spectra exported to 'all_imputedNEW.pdf'.")
print("CSV of final imputed spectra exported to '03-Base-imputedNEW.csv'.")

C:\Users\anaoj\AppData\Local\Temp\ipykernel_24952\1160190029.py:415: UserWarning: Creating legend with loc="best" can be slow with large amounts of data.
  pdf.savefig(fig, bbox_inches='tight')


PDF of all imputed spectra exported to 'all_imputedNEW.pdf'.
CSV of final imputed spectra exported to '03-Base-imputedNEW.csv'.


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Assuming your original DataFrame 'df' is already loaded
# No dummy data generation here. We assume 'df' exists.

# ---------------------------
# Load Data
# ---------------------------

# Assume spectral data are in the first 53 columns.
df_spectra = df.iloc[:, 0:53].copy()
wavelengths = np.array([float(col) for col in df_spectra.columns])

# Ensure the wavelength range is within 0.45 to 2.45
valid_wavelength_indices = np.where((wavelengths >= 0.45) & (wavelengths <= 2.45))[0]
# Check if any valid indices were found
if len(valid_wavelength_indices) == 0:
    raise ValueError("No wavelengths found in the 0.45 to 2.45 range in the first 53 columns.")
wavelengths_subset = wavelengths[valid_wavelength_indices]
# Select the subset columns from the original df_spectra
df_spectra_subset = df_spectra.iloc[:, valid_wavelength_indices].copy()


# ---------------------------
# Load class labels
# ---------------------------
if 'class_asteroid_sf' in df.columns:
    classes = df['class_asteroid_sf']
else:
    raise ValueError("No 'class_asteroid_sf' column found in the CSV data.")

# ---------------------------
# Define the target indices
# ---------------------------
target_indices_to_plot = [1339]  # Removed 1341
slope_weight = 1.0  # You might need to adjust this weight

# ---------------------------
# Function to compute similarity aligned (from your reference)
# ---------------------------
def compute_similarity_aligned(target, candidate, indices, slope_weight, side="left"):
    # Ensure inputs are numpy arrays and indices is not empty
    target = np.asarray(target)
    candidate = np.asarray(candidate)
    indices = np.asarray(indices) # Relative indices within the overlap part

    if indices.size == 0:
        return 1e9 # High error for no overlap

    # Handle case where only one point overlaps
    if indices.size < 2:
        # Calculate MSE based on the single point
        shift_point_single = target[indices[0]] - candidate[indices[0]]
        candidate_aligned_single = candidate + shift_point_single
        # MSE on the single point (after alignment) - could also just be the diff squared
        # Let's calculate MSE based on mean shift even for one point for consistency, though slope is 0
        shift_avg_single = target[indices[0]] - candidate[indices[0]] # Mean shift is just the shift
        candidate_aligned_avg_single = candidate + shift_avg_single
        mse_single = (target[indices[0]] - candidate_aligned_avg_single[indices[0]])**2
        return mse_single # Return only MSE as slope_diff is 0 or undefined

    # --- MSE Calculation (using mean shift) ---
    # Use only the values corresponding to the relative indices
    target_overlap = target[indices]
    candidate_overlap = candidate[indices]

    shift_avg = np.mean(target_overlap - candidate_overlap)
    candidate_aligned_avg_overlap = candidate_overlap + shift_avg
    mse = np.mean((target_overlap - candidate_aligned_avg_overlap)**2)

    # --- Slope Difference Calculation (using point alignment) ---
    # Align based on the first or last point of the *overlap*
    if side == "left":
        # Use relative index 0
        shift_point = target_overlap[0] - candidate_overlap[0]
    else: # side == "right"
        # Use relative index -1
        shift_point = target_overlap[-1] - candidate_overlap[-1]

    candidate_aligned_point_overlap = candidate_overlap + shift_point

    # Calculate slopes on the overlapping region using the point-aligned candidate overlap
    target_slopes = np.diff(target_overlap)
    candidate_slopes = np.diff(candidate_aligned_point_overlap)
    slope_diff = np.mean(np.abs(target_slopes - candidate_slopes))

    total_error = mse + slope_weight * slope_diff
    return total_error if np.isfinite(total_error) else 1e9

# ---------------------------
# Function to compute MSE for alignment (Not directly used in final imputation, but used by find_top_n_candidates_mse)
# ---------------------------
def compute_mse(target, candidate, indices):
    if not indices.size:
        return float('inf')
    # Ensure target and candidate are properly indexed if indices are relative
    return np.mean((target[indices] - candidate[indices])**2)

# ---------------------------
# Function to align candidate based on minimum MSE (for plot 3)
# This function aligns the *part* of the candidate corresponding to 'indices'.
# It should likely return the *shift* or align the whole candidate based on this part.
# Let's modify it slightly to return the aligned *part* for consistency with its use.
# ---------------------------
def align_candidate_mse(target_part, candidate_part, indices_relative):
    """Aligns candidate_part to target_part over relative indices using MSE minimum shift."""
    if not indices_relative.size or target_part.size != candidate_part.size:
        return candidate_part # Return original part if no indices or size mismatch

    target_subset = target_part[indices_relative]
    candidate_subset = candidate_part[indices_relative]

    if target_subset.size == 0: # Check if subset selection resulted in empty array
        return candidate_part

    best_shift = 0
    min_mse = float('inf')

    # Simplified: Calculate optimal shift directly
    valid = ~np.isnan(target_subset) & ~np.isnan(candidate_subset)
    if np.sum(valid) > 0:
         best_shift = np.mean(target_subset[valid]) - np.mean(candidate_subset[valid])
    else:
         best_shift = 0 # No valid points to compare

    # Apply the shift to the original candidate_part passed to the function
    return candidate_part + best_shift


# ---------------------------
# Function to highlight contiguous regions
# ---------------------------
def highlight_contiguous_regions(ax, wavelengths, indices, color, alpha, label):
    # Ensure indices is a numpy array
    indices = np.asarray(indices)
    if indices.size == 0:
        return

    indices = np.sort(indices) # Ensure indices are sorted

    # Find start and end points of contiguous blocks
    diff = np.diff(indices)
    starts = np.concatenate(([indices[0]], indices[1:][diff > 1]))
    ends = np.concatenate((indices[:-1][diff > 1], [indices[-1]]))

    for i, (start_idx, end_idx) in enumerate(zip(starts, ends)):
        current_label = label if i == 0 else None # Label only the first block
        # Safety check for index bounds
        wl_start = wavelengths[start_idx] if start_idx < len(wavelengths) else wavelengths[-1]
        wl_end = wavelengths[end_idx] if end_idx < len(wavelengths) else wavelengths[-1]

        if wl_start <= wl_end:
             ax.axvspan(wl_start, wl_end, color=color, alpha=alpha, label=current_label, zorder=0)


# ---------------------------
# Find top N candidate spectra based on MSE (for plot 3)
# Uses the globally defined 'wavelengths' and 'valid_wavelength_indices'
# Takes df_spectra_subset as input
# ---------------------------
def find_top_n_candidates_mse(df_spectra_subset_local, classes_local, target_index_local, target_spectrum_observed_part, observed_indices_subset_local, nan_indices_subset_local, n=10):
    candidates_with_mse = []
    target_class_local = classes_local.iloc[target_index_local]
    num_observed = len(observed_indices_subset_local)

    if num_observed == 0:
        print(f"Warning (MSE): Target {target_index_local} has no observed points. Cannot find MSE candidates.")
        return []

    # Define the indices that the candidate *must* have non-NaN values for
    # These are the observed indices of the target, and potentially the missing ones if needed later
    # For MSE alignment, we only strictly need the observed overlap non-NaN
    # Let's also check the left missing part needed for *potential* later use or consistency
    first_obs_idx = observed_indices_subset_local[0]
    left_missing_indices_subset_needed = nan_indices_subset_local[nan_indices_subset_local < first_obs_idx]

    for i in range(len(df_spectra_subset_local)):
        # Use iloc index i for df_spectra_subset_local, but check original index 'target_index_local' against df's index
        original_df_index = df_spectra_subset_local.index[i]
        if original_df_index == target_index_local or classes_local.loc[original_df_index] != target_class_local:
            continue

        candidate_subset = df_spectra_subset_local.iloc[i].values.astype(float)

        # Check if candidate has NaNs in the target's observed region or the target's left missing region
        has_nan_in_observed = np.any(np.isnan(candidate_subset[observed_indices_subset_local]))
        has_nan_in_left_missing = np.any(np.isnan(candidate_subset[left_missing_indices_subset_needed]))

        if not has_nan_in_observed and not has_nan_in_left_missing:
            candidate_observed_part = candidate_subset[observed_indices_subset_local]

            # Align the candidate's observed part to the target's observed part
            # align_candidate_mse now returns the *aligned candidate part*
            aligned_candidate_part = align_candidate_mse(target_spectrum_observed_part, candidate_observed_part, np.arange(num_observed))

            # Calculate MSE between target observed part and the *aligned* candidate part
            mse = np.mean((target_spectrum_observed_part - aligned_candidate_part)**2)

            if np.isfinite(mse):
                # Store the original candidate subset spectrum and its alignment MSE
                candidates_with_mse.append((candidate_subset, mse)) # Store tuple (spectrum, mse)

    candidates_with_mse.sort(key=lambda item: item[1]) # Sort by MSE (second element)
    # Return only the spectra of the top N candidates
    top_candidates_spectra = [item[0] for item in candidates_with_mse[:n]]
    return top_candidates_spectra


# ---------------------------
# Find top N candidate spectra based on similarity (MSE + Slope)
# Uses the globally defined 'wavelengths' and 'valid_wavelength_indices'
# Takes df_spectra_subset as input
# **MODIFIED TO RETURN LIST OF (spectrum, error) TUPLES**
# ---------------------------
def find_top_n_candidates(df_spectra_subset_local, classes_local, target_index_local, target_spectrum_observed_part, observed_indices_subset_local, nan_indices_subset_local, slope_weight_local, n=10, side="left"):
    candidates_with_error = []
    target_class_local = classes_local.iloc[target_index_local]
    num_observed = len(observed_indices_subset_local)

    # Need at least 2 observed points to calculate slope difference within compute_similarity_aligned
    if num_observed < 2:
        print(f"Warning (Similarity): Target {target_index_local} has < 2 observed points. Cannot reliably calculate similarity including slope.")
        # Return empty list as the similarity metric relies on slope diff
        return []

    # Define the indices that the candidate *must* have non-NaN values for
    first_obs_idx = observed_indices_subset_local[0]
    left_missing_indices_subset_needed = nan_indices_subset_local[nan_indices_subset_local < first_obs_idx]

    for i in range(len(df_spectra_subset_local)):
        original_df_index = df_spectra_subset_local.index[i]
        if original_df_index == target_index_local or classes_local.loc[original_df_index] != target_class_local:
            continue

        candidate_subset = df_spectra_subset_local.iloc[i].values.astype(float)

        # Check if candidate has NaNs in the target's observed region or the target's left missing region
        has_nan_in_observed = np.any(np.isnan(candidate_subset[observed_indices_subset_local]))
        has_nan_in_left_missing = np.any(np.isnan(candidate_subset[left_missing_indices_subset_needed]))

        if not has_nan_in_observed and not has_nan_in_left_missing:
            candidate_observed_part = candidate_subset[observed_indices_subset_local]

            # Compute similarity using the observed parts and relative indices
            error = compute_similarity_aligned(target_spectrum_observed_part, candidate_observed_part,
                                               np.arange(num_observed), # Relative indices for the overlap part
                                               slope_weight_local, side=side)

            if np.isfinite(error):
                 candidates_with_error.append((candidate_subset, error)) # Store tuple (spectrum, error)


    candidates_with_error.sort(key=lambda item: item[1]) # Sort by error (second element)

    # *** MODIFICATION: Return the list of tuples (spectrum, error) ***
    return candidates_with_error[:n]


# ---------------------------
# Function to align a candidate (using point alignment for plotting)
# Aligns the *entire* candidate spectrum based on one point in the overlap.
# ---------------------------
def align_candidate_point(target_subset, candidate_subset, anchor_indices_subset, side="left"):
    """Aligns the full candidate_subset based on an anchor point comparison."""
    if not anchor_indices_subset.size:
        return candidate_subset # No anchor point provided

    if side == "left":
        anchor_idx = anchor_indices_subset[0]
    else: # side == "right"
        anchor_idx = anchor_indices_subset[-1]

    # Check if anchor point is valid in both target and candidate
    if anchor_idx >= len(target_subset) or anchor_idx >= len(candidate_subset) or \
       np.isnan(target_subset[anchor_idx]) or np.isnan(candidate_subset[anchor_idx]):
        # print(f"Warning: Cannot align using anchor index {anchor_idx}. Values invalid or out of bounds.")
        return candidate_subset # Return unaligned if anchor is bad

    shift = target_subset[anchor_idx] - candidate_subset[anchor_idx]
    return candidate_subset + shift


# ---------------------------
# Plot the specified target spectra
# ---------------------------
print(f"Starting analysis for target indices: {target_indices_to_plot}")
for target_index in target_indices_to_plot:
    # Check if target index exists in the original DataFrame index
    if target_index not in df.index:
         print(f"Index {target_index} is out of bounds or not found in the DataFrame index. Skipping.")
         continue

    # Get data using the subset DataFrame for spectra, but original index for class
    # Use .loc to safely access row by index label from df_spectra_subset
    if target_index not in df_spectra_subset.index:
         print(f"Index {target_index} not found in df_spectra_subset (possibly filtered out by wavelength range). Skipping.")
         continue

    sample_subset = df_spectra_subset.loc[target_index].values.astype(float)
    target_class = classes.loc[target_index] # Use original index for classes Series

    print(f"\nProcessing Target Index: {target_index} (Class: {target_class})")

    # Check if it belongs to class 'Q' and has NaNs within the subset
    if target_class == 'Q' and np.any(np.isnan(sample_subset)):
        # ---------------------------
        # Identify missing and observed indices (relative to sample_subset/wavelengths_subset)
        # ---------------------------
        nan_indices_subset = np.where(np.isnan(sample_subset))[0]
        observed_indices_subset = np.where(~np.isnan(sample_subset))[0]

        if observed_indices_subset.size == 0:
             print(f"Skipping index {target_index}: No observed data points in the 0.45-2.45 µm range.")
             continue

        print(f"  Observed indices count: {len(observed_indices_subset)}, NaN indices count: {len(nan_indices_subset)}")

        # ---------------------------
        # Find Top Candidates
        # ---------------------------

        # --- Find top 10 candidates based on SIMILARITY (for imputation & plot 2) ---
        # Note: find_top_n_candidates now returns list of (spectrum, error) tuples
        top_candidates_left_info = []
        target_observed_part = sample_subset[observed_indices_subset] # Get the observed part of the target
        # Check if enough observed points for similarity calculation
        if observed_indices_subset.size >= 2:
             top_candidates_left_info = find_top_n_candidates(
                 df_spectra_subset, classes, target_index, # Pass necessary args
                 target_observed_part, # Pass only the observed part for comparison
                 observed_indices_subset, nan_indices_subset,
                 slope_weight, n=10, side="left"
             )
             print(f"  Found {len(top_candidates_left_info)} candidates based on similarity.")
        else:
             print(f"  Skipping similarity candidates: Need >= 2 observed points (found {observed_indices_subset.size}).")


        # --- Find top 10 candidates based on MSE alignment (for plot 3) ---
        # Note: find_top_n_candidates_mse returns only spectra list
        top_candidates_mse_aligned_spectra = []
        if observed_indices_subset.size >= 1: # Need at least 1 point for MSE alignment
            top_candidates_mse_aligned_spectra = find_top_n_candidates_mse(
                 df_spectra_subset, classes, target_index, # Pass necessary args
                 target_observed_part, # Pass only the observed part
                 observed_indices_subset, nan_indices_subset, n=10
             )
            print(f"  Found {len(top_candidates_mse_aligned_spectra)} candidates based on MSE.")
        else:
             print(f"  Skipping MSE candidates: Need >= 1 observed point.")


        # ---------------------------
        # Create the figure with four subplots in a 2x2 grid
        # ---------------------------
        fig, axes = plt.subplots(2, 2, figsize=(20, 10), sharey=True) # Adjusted figsize slightly
        fig.suptitle(f"Analysis for Target Spectrum Index {target_index} (Class {target_class})", fontsize=24, y=1.02)
        common_xlim = (wavelengths_subset[0] - 0.05, wavelengths_subset[-1] + 0.05)


        # --- Plot on the first subplot (axes[0, 0]) ---
        ax1 = axes[0, 0]
        ax1.plot(wavelengths_subset, sample_subset, 'ko-', label='Target Spectrum', markersize=5, linewidth=1.2)
        highlight_contiguous_regions(ax1, wavelengths_subset, nan_indices_subset, 'red', 0.2, 'Missing Region')
        highlight_contiguous_regions(ax1, wavelengths_subset, observed_indices_subset, 'green', 0.2, 'Observed Region')
        ax1.set_xlabel("Wavelength (µm)", fontsize=24)
        ax1.set_ylabel("Reflectance", fontsize=24)
        ax1.legend(fontsize=19)
        ax1.tick_params(axis='both', which='major', labelsize=19)
        ax1.set_xlim(common_xlim)

        # --- Plot on the second subplot (axes[0, 1]) with left-aligned candidates (from similarity) ---
        ax2 = axes[0, 1]
        ax2.plot(wavelengths_subset, sample_subset, 'ko-', label='Target Spectrum', markersize=5, linewidth=1.2, zorder=10)
        #highlight_contiguous_regions(ax2, wavelengths_subset, nan_indices_subset, 'red', 0.2, '_nolegend_')
        highlight_contiguous_regions(ax2, wavelengths_subset, observed_indices_subset, 'green', 0.2, '_nolegend_')
        ax2.set_xlabel("Wavelength (µm)", fontsize=24)
        ax2.tick_params(axis='both', which='major', labelsize=19)
        ax2.yaxis.set_visible(False)  # Hide y-axis labels and ticks

        # Align candidates visually based on the first observed point
        anchor_plot_indices = observed_indices_subset[:1] # Use first observed point as anchor for plot 2
        if anchor_plot_indices.size > 0 and top_candidates_left_info:
            # Extract just the spectra for plotting
            top_spectra_for_plot2 = [info[0] for info in top_candidates_left_info]
            for candidate_spec in top_spectra_for_plot2:
                 # Align the *entire* candidate spectrum for visualization
                 aligned_candidate_vis = align_candidate_point(sample_subset, candidate_spec, anchor_plot_indices, side="left")
                 ax2.plot(wavelengths_subset, aligned_candidate_vis, color='gray', linewidth=0.7, alpha=0.6, zorder=1)
        ax2.set_xlim(common_xlim)

        # --- Plot on the third subplot (axes[1, 0]) with MSE-aligned candidates ---
        ax3 = axes[1, 0]
        ax3.plot(wavelengths_subset, sample_subset, 'ko-', label='Target Spectrum', markersize=5, linewidth=1.2, zorder=10)
        #highlight_contiguous_regions(ax3, wavelengths_subset, nan_indices_subset, 'red', 0.2, '_nolegend_')
        highlight_contiguous_regions(ax3, wavelengths_subset, observed_indices_subset, 'green', 0.2, '_nolegend_')
        ax3.set_xlabel("Wavelength (µm)", fontsize=24)
        ax3.set_ylabel("Reflectance", fontsize=24) # Show Y label here again
        ax3.tick_params(axis='both', which='major', labelsize=19)

        if top_candidates_mse_aligned_spectra and observed_indices_subset.size > 0:
            target_obs_part_for_mse = sample_subset[observed_indices_subset]
            for candidate_spec in top_candidates_mse_aligned_spectra:
                candidate_obs_part = candidate_spec[observed_indices_subset]
                # Calculate the optimal shift based on MSE over the observed part
                best_shift_mse = align_candidate_mse(target_obs_part_for_mse, candidate_obs_part, np.arange(len(target_obs_part_for_mse))) # align_candidate_mse returns shift now? No, returns aligned part. Let's recalculate shift
                valid_mse = ~np.isnan(target_obs_part_for_mse) & ~np.isnan(candidate_obs_part)
                if np.sum(valid_mse) > 0:
                     best_shift_mse_val = np.mean(target_obs_part_for_mse[valid_mse]) - np.mean(candidate_obs_part[valid_mse])
                else: best_shift_mse_val = 0

                # Apply shift to the *entire* candidate spectrum for plotting
                aligned_candidate_mse_plot = candidate_spec + best_shift_mse_val
                ax3.plot(wavelengths_subset, aligned_candidate_mse_plot, color='grey', linewidth=0.7, alpha=0.8, zorder=1) # Use lightblue for MSE plot
        ax3.set_xlim(common_xlim)

        # ********************************************************************
        # --- Calculate Imputed Data & Plot on Fourth Subplot (axes[1, 1]) ---
        # ********************************************************************
        imputed_values_plot = np.array([]) # Initialize empty array for imputed points
        left_missing_indices_subset = np.array([], dtype=int) # Initialize

        # Use candidates found by similarity metric (top_candidates_left_info)
        # Need candidates with errors, observed points, and missing points to the left
        if top_candidates_left_info and observed_indices_subset.size > 0:
            # Identify indices within the 'subset' that are NaN AND before the first observed point
            first_observed_idx_in_subset = observed_indices_subset[0]
            left_missing_indices_subset = nan_indices_subset[nan_indices_subset < first_observed_idx_in_subset]

            # Define overlap indices for alignment/slope (use first few observed points)
            # Need at least 2 points for slope calculation for the smoothing step
            num_overlap_points_impute = min(5, len(observed_indices_subset)) # Use up to 5 points
            left_overlap_indices_subset = observed_indices_subset[:num_overlap_points_impute]

            if left_missing_indices_subset.size > 0 and left_overlap_indices_subset.size > 0:
                print(f"Target {target_index}: Calculating imputation for {len(left_missing_indices_subset)} left missing points.")

                # Extract candidate spectra and errors from the list of tuples
                imputed_left_candidates_spectra = [info[0] for info in top_candidates_left_info]
                errors_left = np.array([info[1] for info in top_candidates_left_info])

                imputed_left_values_collected = [] # To store the relevant part from each *aligned* candidate

                # Use the first overlap point as the reference for alignment shift for imputation
                anchor_idx_impute_subset = left_overlap_indices_subset[0]

                for cand_spec_subset in imputed_left_candidates_spectra:
                    # Align the candidate based on the anchor point
                    # Ensure anchor point is valid before aligning
                    shift_point_impute = 0
                    if anchor_idx_impute_subset < len(sample_subset) and anchor_idx_impute_subset < len(cand_spec_subset) and \
                       not np.isnan(sample_subset[anchor_idx_impute_subset]) and not np.isnan(cand_spec_subset[anchor_idx_impute_subset]):
                         shift_point_impute = sample_subset[anchor_idx_impute_subset] - cand_spec_subset[anchor_idx_impute_subset]
                    else:
                        # Handle case where anchor point itself is NaN - maybe skip this candidate or use mean shift?
                        # For now, let's assume a zero shift if anchor is bad, though this might be suboptimal
                        print(f"Warning: Anchor point {anchor_idx_impute_subset} invalid for alignment in imputation. Using shift=0 for this candidate.")

                    cand_aligned = cand_spec_subset + shift_point_impute
                    # Extract the values corresponding to the *left missing* indices from the aligned candidate
                    imputed_left_values_collected.append(cand_aligned[left_missing_indices_subset])

                if imputed_left_values_collected: # Ensure we collected some values
                    imputed_left_values_collected = np.array(imputed_left_values_collected)

                    # --- Weighted Average ---
                    weights_left = 1.0 / (errors_left + 1e-9) # Add epsilon for stability
                    # Check if weights are valid before averaging
                    if np.sum(weights_left) > 0 and np.all(np.isfinite(weights_left)):
                        weighted_imputed_left = np.average(imputed_left_values_collected, axis=0, weights=weights_left)
                    else:
                         print(f"Target {target_index}: Warning - Invalid weights for imputation. Using simple mean.")
                         # Ensure errors_left and imputed_left_values_collected match in number of candidates
                         if len(errors_left) == imputed_left_values_collected.shape[0]:
                             weighted_imputed_left = np.mean(imputed_left_values_collected, axis=0)
                         else:
                              print(f"Target {target_index}: Error - Mismatch in number of errors and collected values. Cannot compute mean.")
                              weighted_imputed_left = np.full(left_missing_indices_subset.shape, np.nan) # Fallback


                    # --- Smoothing/Extrapolation based on slope (optional, based on snippet) ---
                    if left_overlap_indices_subset.size >= 2:
                        # Calculate slope from the first two observed points of the target
                        idx0 = left_overlap_indices_subset[0]
                        idx1 = left_overlap_indices_subset[1]
                        x0_wl = wavelengths_subset[idx0]
                        x1_wl = wavelengths_subset[idx1]
                        y0_val = sample_subset[idx0]
                        y1_val = sample_subset[idx1]
                        # Avoid division by zero if wavelengths are identical
                        slope_left = (y1_val - y0_val) / (x1_wl - x0_wl) if (x1_wl - x0_wl) != 0 else 0
                    else:
                        slope_left = 0 # Cannot calculate slope with < 2 points

                    # Apply smoothing using distance-based weights
                    smoothed_left = weighted_imputed_left.copy()
                    first_obs_idx_for_smoothing = observed_indices_subset[0] # Index within the subset

                    for idx_in_smoothed_array, missing_idx_in_subset in enumerate(left_missing_indices_subset):
                        d = first_obs_idx_for_smoothing - missing_idx_in_subset # Distance in indices

                        # Define weights based on distance 'd' (same as snippet)
                        if d == 1: w = 0.5
                        elif d == 2: w = 0.35
                        elif d == 3: w = 0.25
                        elif d == 4: w = 0.15
                        elif d == 5: w = 0.10
                        elif d == 6: w = 0.05
                        else: w = 0.0 # No smoothing for points far away

                        if w > 0 and slope_left != 0 and np.isfinite(slope_left):
                            # Extrapolate from the first observed point
                            missing_wl = wavelengths_subset[missing_idx_in_subset]
                            x0_wl_smooth = wavelengths_subset[first_obs_idx_for_smoothing] # Wavelength of first observed
                            y0_val_smooth = sample_subset[first_obs_idx_for_smoothing] # Value of first observed

                            # Ensure y0_val_smooth is valid before extrapolating
                            if not np.isnan(y0_val_smooth):
                                 extrapolated_val = y0_val_smooth - slope_left * (x0_wl_smooth - missing_wl) # y = y0 + m*(x-x0) -> y = y0 - m*(x0-x)
                                 # Combine weighted average with extrapolation, check if current value is valid
                                 current_weighted_val = weighted_imputed_left[idx_in_smoothed_array]
                                 if np.isfinite(extrapolated_val) and np.isfinite(current_weighted_val):
                                      smoothed_left[idx_in_smoothed_array] = w * extrapolated_val + (1 - w) * current_weighted_val
                                 elif np.isfinite(extrapolated_val): # If only extrapolation is valid, use it partially? Or just keep original?
                                     # Maybe lean towards extrapolation if weighted avg failed? Or stick to weighted if valid?
                                     # Let's stick to weighted if valid, otherwise try extrapolation biased?
                                     # Safest: only combine if both are valid.
                                     pass # Keep original weighted if extrapolation or weighted is NaN/inf
                            # else: Cannot extrapolate if the anchor point is NaN

                        # else: keep the original weighted_imputed_left[idx_in_smoothed_array] value if w=0 or slope invalid

                    # Final imputed values for plotting are the smoothed ones
                    imputed_values_plot = smoothed_left
                    print(f"Target {target_index}: Imputation & Smoothing successful.")
                else:
                     print(f"Target {target_index}: Imputation failed: No candidate values were collected after alignment.")
            else:
                if left_missing_indices_subset.size == 0:
                    print(f"Target {target_index}: Skipping imputation: No missing values found to the left of observed data.")
                elif left_overlap_indices_subset.size == 0:
                    print(f"Target {target_index}: Skipping imputation: No overlap points found.")
                # Handle case where top_candidates_left_info was empty
                elif not top_candidates_left_info:
                     print(f"Target {target_index}: Skipping imputation: No similarity candidates found.")


        # --- Plot on the fourth subplot (axes[1, 1]) ---
        ax4 = axes[1, 1]
        ax4.plot(wavelengths_subset, sample_subset, 'ko-', label='Target Spectrum', markersize=5, zorder=10, linewidth=1.2) # Reduced marker size
        # Highlight regions again for context
        highlight_contiguous_regions(ax4, wavelengths_subset, nan_indices_subset, 'red', 0.2, 'Missing Region')
        #highlight_contiguous_regions(ax4, wavelengths_subset, observed_indices_subset, 'green', 0.2, 'Observed Region')

        # Plot the imputed points if they exist and match the number of missing indices
        if imputed_values_plot.size > 0 and left_missing_indices_subset.size == imputed_values_plot.size :
            ax4.plot(wavelengths_subset[left_missing_indices_subset], imputed_values_plot,
                     'ro', # Red dots ('o')
                     label='Imputed Data',
                     markersize=6, # Adjust marker size as needed
                     zorder=20) # Ensure points are plotted on top
        elif imputed_values_plot.size > 0:
             print(f"Target {target_index}: Warning - Mismatch between number of imputed points ({imputed_values_plot.size}) and missing indices ({left_missing_indices_subset.size}). Skipping plot.")


        ax4.set_xlabel("Wavelength (µm)", fontsize=24)
        ax4.tick_params(axis='x', labelsize=19)
        ax4.yaxis.set_visible(False)  # Hide y-axis labels and ticks
        ax4.legend(fontsize=19) # Added legend
        ax4.set_xlim(common_xlim) # Use common xlim

        # Align candidates visually based on the first observed point
        anchor_plot_indices = observed_indices_subset[:1] # Use first observed point as anchor for plot 2
        if anchor_plot_indices.size > 0 and top_candidates_left_info:
            # Extract just the spectra for plotting
            top_spectra_for_plot2 = [info[0] for info in top_candidates_left_info]
            for candidate_spec in top_spectra_for_plot2:
                 # Align the *entire* candidate spectrum for visualization
                 aligned_candidate_vis = align_candidate_point(sample_subset, candidate_spec, anchor_plot_indices, side="left")
                 ax4.plot(wavelengths_subset, aligned_candidate_vis, color='gray', linewidth=0.7, alpha=0.6, zorder=1)
        ax4.set_xlim(common_xlim)

        
        # --- Final Figure Adjustments ---
        plt.tight_layout(rect=[0, 0.03, 1, 0.97]) # Adjust layout slightly
        plt.show()

    # --- End of Q class check ---
    elif target_class != 'Q':
        print(f"Skipping index {target_index}: Class is '{target_class}', not 'Q'.")
    else: # Class is Q, but no NaNs in the subset
        print(f"Skipping index {target_index}: Class 'Q' spectrum has no missing values in the 0.45-2.45 µm range.")

# --- End of loop ---
print("\nProcessing complete.")

## Final method - Albedo

In [162]:
import pandas as pd

# Identify missing pV values
missing_pV_data = df[df['pV'].isna()]
names_missing_pV = missing_pV_data['name'].unique()

# Fill missing values with mean pV of the same name
for name in names_missing_pV:
#    print(f"Name: {name} (Missing pV data)")
    all_entries_for_name = df[df['name'] == name]
    available_pV_data = all_entries_for_name[all_entries_for_name['pV'].notna()]
    
    if not available_pV_data.empty:
        mean_pV = available_pV_data['pV'].mean()
        df.loc[(df['name'] == name) & (df['pV'].isna()), 'pV'] = mean_pV
#        print(f"Filled missing 'pV' for {name} with mean value: {mean_pV:.4f}")
#    else:
#        print("No entries with available pV data.")
#    print("\n" + "-"*50 + "\n")

# Save the updated DataFrame with the same original columns
df.to_csv('03-Base-imputedNew3.csv', index=False)
print("Missing 'pV' values have been filled with mean values, and the updated CSV file '03-Base-imputedNew3.csv' has been saved.")

Missing 'pV' values have been filled with mean values, and the updated CSV file '03-Base-imputedNew3.csv' has been saved.


In [164]:
# Load the updated CSV and count remaining missing pV values
updated_df = pd.read_csv('03-Base-imputedNew3.csv')
remaining_missing_pV = updated_df['pV'].isna().sum()
print(f"Total remaining missing 'pV' values: {remaining_missing_pV}")

Total remaining missing 'pV' values: 551


In [166]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages

# Set global font to DejaVu Serif and increase font sizes
plt.rcParams['font.family'] = 'DejaVu Serif'
plt.rcParams['font.size'] = 20

# Load CSV
csv_file_path = '03-Base-imputedNew3.csv'
df = pd.read_csv(csv_file_path)

# Identify remaining missing pV values
remaining_missing_pV_data = df[df['pV'].isna()]
remaining_missing_pV_count = remaining_missing_pV_data.shape[0]
print(f"Total remaining missing 'pV' values: {remaining_missing_pV_count}")

# Fill missing pV values based on class
pdf_filename = 'pV_class_analysis.pdf'
with PdfPages(pdf_filename) as pdf:
    for classi in df['class_asteroid_sf'].unique():
        class_samples = df[df['class_asteroid_sf'] == classi]
        pV_values_class = class_samples['pV'].dropna()
        
        if not pV_values_class.empty:
            Q1 = pV_values_class.quantile(0.25)
            Q2 = pV_values_class.quantile(0.50)  # Median (Q2)
            Q3 = pV_values_class.quantile(0.75)
            IQR = Q3 - Q1
            lower_bound = Q1 - 1.5 * IQR
            upper_bound = Q3 + 1.5 * IQR
            
            is_outlier = (pV_values_class < lower_bound) | (pV_values_class > upper_bound)
            weights = np.where(is_outlier, 0.1, 1.0)
            weighted_mean_pV_class = np.average(pV_values_class, weights=weights)
            
            missing_pV_indices = (df['class_asteroid_sf'] == classi) & (df['pV'].isna())
            df.loc[missing_pV_indices, 'pV'] = weighted_mean_pV_class
            
            # Generate plot
            plt.figure(figsize=(10, 4))  # Decreased height to shrink the plot on y-axis
            box = plt.boxplot(pV_values_class, vert=False, patch_artist=True, showfliers=True, 
                              flierprops=dict(marker='o', color='red', alpha=0.5), widths=0.3)  # Further shrinking boxplot height
            
            # Increase thickness of Q2 (median) line
            for median in box['medians']:
                median.set(linewidth=3.5, color='orange')
            
            plt.axvline(x=lower_bound, color='blue', linestyle='--', label=f'Lower bound (Outlier): {lower_bound:.2f}', linewidth=3)
            plt.axvline(x=upper_bound, color='blue', linestyle='--', label=f'Upper bound (Outlier): {upper_bound:.2f}', linewidth=3)
            plt.title(f'Box plot of "pV" for class {classi} with outliers')
            plt.xlabel('log10 pV')
            plt.legend(fontsize=12)
            plt.tight_layout(rect=[0, 0.02, 1, 1])  # Adjust layout to move content up
            pdf.savefig()
            plt.close()

print(f"The PDF file with results and plots has been saved as '{pdf_filename}'.")

# Save the updated DataFrame with filled pV values
df.to_csv('03-Base-imputedNew223.csv', index=False)
print("Update complete. Missing 'pV' values have been imputed.")

Total remaining missing 'pV' values: 551
The PDF file with results and plots has been saved as 'pV_class_analysis.pdf'.
Update complete. Missing 'pV' values have been imputed.


## Final database (concatenated)

In [168]:
import pandas as pd
df1 = pd.read_csv('03-Base-imputedNew.csv')  # Original spectra dataset
df2 = pd.read_csv('03-Base-imputedNew223.csv')  # Contains extra columns
extra_columns = ['pV', 'name', 'counts', 'class_bdm', 'class_asteroid_sf']
df2_extra = df2[extra_columns]
# Ensure both datasets have the same length before concatenation
if len(df1) == len(df2_extra):
    merged_df = pd.concat([df1, df2_extra], axis=1)  # Concatenate columns
    merged_df.to_csv('05-BaseNew.csv', index=False)
    print("Merging complete. The new file '05-BaseNew.csv' has been created.")
else:
    print("Error: The two datasets have different numbers of rows. Check for missing or extra data.")

Merging complete. The new file '05-BaseNew.csv' has been created.


In [170]:
merged_df

0.45     0.475       0.5     0.525      0.55     0.575       0.6  \
0     0.474660  0.510748  0.547428  0.588601  0.620076  0.656002  0.683170   
1     0.405455  0.433723  0.469353  0.504947  0.536460  0.567352  0.593231   
2     0.317047  0.347344  0.384102  0.419826  0.452776  0.484209  0.509688   
3     0.479208  0.514880  0.554585  0.594870  0.628223  0.663880  0.689781   
4     0.399207  0.429273  0.465856  0.502021  0.535473  0.567403  0.593125   
...        ...       ...       ...       ...       ...       ...       ...   
2935  0.433300  0.449156  0.464809  0.479891  0.489143  0.503325  0.517542   
2936  0.467633  0.480760  0.494760  0.508886  0.516543  0.529372  0.542561   
2937  0.424824  0.438244  0.452745  0.467561  0.482917  0.497040  0.508614   
2938  0.654094  0.649106  0.647750  0.649133  0.649091  0.652669  0.654123   
2939  0.520102  0.538432  0.554007  0.567428  0.579296  0.590139  0.601631   

         0.625      0.65     0.675  ...      2.25       2.3      2.35  \
0     0.709885  0.741209  0.772798  ...  1.195193  1.196479  1.194755   
1     0.622140  0.650964  0.675510  ...  1.260412  1.261258  1.244087   
2     0.536324  0.569470  0.597419  ...  1.246545  1.253701  1.248591   
3     0.716422  0.756680  0.787789  ...  1.167623  1.163372  1.155753   
4     0.619699  0.649162  0.676829  ...  1.234882  1.222925  1.222321   
...        ...       ...       ...  ...       ...       ...       ...   
2935  0.531750  0.545955  0.560139  ...  1.378472  1.401883  1.418678   
2936  0.555911  0.570074  0.584662  ...  1.322165  1.346740  1.371653   
2937  0.521423  0.536404  0.551251  ...  1.242833  1.252417  1.266928   
2938  0.661425  0.668186  0.674114  ...  1.272509  1.285464  1.299044   
2939  0.612858  0.624339  0.636803  ...  1.336246  1.353193  1.370389   

           2.4      2.45       pV        name  counts  class_bdm  \
0     1.189291  1.183298  0.25645     1988 TA       1          A   
1     1.238516  1.238515  0.40600    1990 WO3       1          A   
2     1.248602  1.271996  0.25645    1993 FO6       1          A   
3     1.148710  1.146478  0.22100  1998 FE118       1         Sa   
4     1.224672  1.228380  0.37800    1998 YG8       1          A   
...        ...       ...      ...         ...     ...        ...   
2935  1.437153  1.456324  0.22000     Skorina       3          D   
2936  1.394933  1.413759  0.11000     Skorina       1          D   
2937  1.286292  1.317274  0.06700     Thestor       1          D   
2938  1.310190  1.315834  0.04400       Thule       1          D   
2939  1.388239  1.401814  0.05500      Tornio       1          D   

      class_asteroid_sf  
0                     A  
1                     A  
2                     A  
3                     A  
4                     A  
...                 ...  
2935                  Z  
2936                  Z  
2937                  Z  
2938                  Z  
2939                  Z  

[2940 rows x 58 columns]

In [186]:
import pandas as pd

# Load the data
df = pd.read_csv('05-BaseNew.csv')

# Identify the column for 0.55 micrometers (5th column, index 4)
reference_column_index = 4

# Select the columns to normalize (all columns except the last 5)
columns_to_normalize = df.columns[:-5]

# Create a new DataFrame for the normalized data
df_normalized = df[columns_to_normalize].copy()

# Get the reflectance value at 0.55 micrometers for each sample
reference_reflectance = df.iloc[:, reference_column_index]

# Normalize each column by the reference reflectance
for col in columns_to_normalize:
    df_normalized[col] = df_normalized[col] / reference_reflectance

# Add back any columns that were excluded (the last 5)
df_normalized = pd.concat([df_normalized, df[df.columns[-5:]]], axis=1)

# Export the normalized DataFrame to a new CSV file
df_normalized.to_csv('05-BaseNew1.csv', index=False)

print("Data normalized to 0.55 micrometers and saved to '05-BaseNew1.csv'")

Data normalized to 0.55 micrometers and saved to '05-BaseNew1.csv'


In [188]:
df = pd.read_csv('05-BaseNew1.csv')
df

0.45     0.475       0.5     0.525  0.55     0.575       0.6  \
0     0.765487  0.823687  0.882840  0.949241   1.0  1.057939  1.101752   
1     0.755798  0.808492  0.874908  0.941258   1.0  1.057586  1.105827   
2     0.700229  0.767144  0.848328  0.927227   1.0  1.069424  1.125698   
3     0.762798  0.819581  0.882783  0.946908   1.0  1.056758  1.097987   
4     0.745522  0.801671  0.869989  0.937528   1.0  1.059629  1.107666   
...        ...       ...       ...       ...   ...       ...       ...   
2935  0.885836  0.918252  0.950252  0.981087   1.0  1.028994  1.058060   
2936  0.905312  0.930726  0.957829  0.985176   1.0  1.024837  1.050369   
2937  0.879704  0.907492  0.937520  0.968200   1.0  1.029245  1.053212   
2938  1.007708  1.000023  0.997933  1.000065   1.0  1.005513  1.007752   
2939  0.897818  0.929460  0.956346  0.979514   1.0  1.018717  1.038556   

         0.625      0.65     0.675  ...      2.25       2.3      2.35  \
0     1.144836  1.195353  1.246295  ...  1.927495  1.929570  1.926789   
1     1.159715  1.213445  1.259200  ...  2.349500  2.351078  2.319069   
2     1.184525  1.257731  1.319459  ...  2.753119  2.768924  2.757638   
3     1.140394  1.204475  1.253995  ...  1.858611  1.851845  1.839717   
4     1.157293  1.212315  1.263984  ...  2.306153  2.283822  2.282694   
...        ...       ...       ...  ...       ...       ...       ...   
2935  1.087105  1.116146  1.145145  ...  2.818139  2.866001  2.900336   
2936  1.076213  1.103633  1.131874  ...  2.559641  2.607216  2.655446   
2937  1.079734  1.110757  1.141502  ...  2.573592  2.593439  2.623487   
2938  1.019002  1.029418  1.038550  ...  1.960448  1.980407  2.001328   
2939  1.057935  1.077755  1.099271  ...  2.306673  2.335927  2.365611   

           2.4      2.45       pV        name  counts  class_bdm  \
0     1.917978  1.908312  0.25645     1988 TA       1          A   
1     2.308685  2.308683  0.40600    1990 WO3       1          A   
2     2.757661  2.809330  0.25645    1993 FO6       1          A   
3     1.828506  1.824952  0.22100  1998 FE118       1         Sa   
4     2.287084  2.294010  0.37800    1998 YG8       1          A   
...        ...       ...      ...         ...     ...        ...   
2935  2.938107  2.977300  0.22000     Skorina       3          D   
2936  2.700515  2.736961  0.11000     Skorina       1          D   
2937  2.663585  2.727741  0.06700     Thestor       1          D   
2938  2.018500  2.027195  0.04400       Thule       1          D   
2939  2.396425  2.419857  0.05500      Tornio       1          D   

      class_asteroid_sf  
0                     A  
1                     A  
2                     A  
3                     A  
4                     A  
...                 ...  
2935                  Z  
2936                  Z  
2937                  Z  
2938                  Z  
2939                  Z  

[2940 rows x 58 columns]

In [190]:
import pandas as pd
df = pd.read_csv('05-BaseNew1.csv')
column_to_drop = df.columns[4]
df.drop(columns=[column_to_drop], inplace=True)
df

0.45     0.475       0.5     0.525     0.575       0.6     0.625  \
0     0.765487  0.823687  0.882840  0.949241  1.057939  1.101752  1.144836   
1     0.755798  0.808492  0.874908  0.941258  1.057586  1.105827  1.159715   
2     0.700229  0.767144  0.848328  0.927227  1.069424  1.125698  1.184525   
3     0.762798  0.819581  0.882783  0.946908  1.056758  1.097987  1.140394   
4     0.745522  0.801671  0.869989  0.937528  1.059629  1.107666  1.157293   
...        ...       ...       ...       ...       ...       ...       ...   
2935  0.885836  0.918252  0.950252  0.981087  1.028994  1.058060  1.087105   
2936  0.905312  0.930726  0.957829  0.985176  1.024837  1.050369  1.076213   
2937  0.879704  0.907492  0.937520  0.968200  1.029245  1.053212  1.079734   
2938  1.007708  1.000023  0.997933  1.000065  1.005513  1.007752  1.019002   
2939  0.897818  0.929460  0.956346  0.979514  1.018717  1.038556  1.057935   

          0.65     0.675       0.7  ...      2.25       2.3      2.35  \
0     1.195353  1.246295  1.277057  ...  1.927495  1.929570  1.926789   
1     1.213445  1.259200  1.289194  ...  2.349500  2.351078  2.319069   
2     1.257731  1.319459  1.360667  ...  2.753119  2.768924  2.757638   
3     1.204475  1.253995  1.281903  ...  1.858611  1.851845  1.839717   
4     1.212315  1.263984  1.316203  ...  2.306153  2.283822  2.282694   
...        ...       ...       ...  ...       ...       ...       ...   
2935  1.116146  1.145145  1.169446  ...  2.818139  2.866001  2.900336   
2936  1.103633  1.131874  1.161987  ...  2.559641  2.607216  2.655446   
2937  1.110757  1.141502  1.165563  ...  2.573592  2.593439  2.623487   
2938  1.029418  1.038550  1.051106  ...  1.960448  1.980407  2.001328   
2939  1.077755  1.099271  1.118172  ...  2.306673  2.335927  2.365611   

           2.4      2.45       pV        name  counts  class_bdm  \
0     1.917978  1.908312  0.25645     1988 TA       1          A   
1     2.308685  2.308683  0.40600    1990 WO3       1          A   
2     2.757661  2.809330  0.25645    1993 FO6       1          A   
3     1.828506  1.824952  0.22100  1998 FE118       1         Sa   
4     2.287084  2.294010  0.37800    1998 YG8       1          A   
...        ...       ...      ...         ...     ...        ...   
2935  2.938107  2.977300  0.22000     Skorina       3          D   
2936  2.700515  2.736961  0.11000     Skorina       1          D   
2937  2.663585  2.727741  0.06700     Thestor       1          D   
2938  2.018500  2.027195  0.04400       Thule       1          D   
2939  2.396425  2.419857  0.05500      Tornio       1          D   

      class_asteroid_sf  
0                     A  
1                     A  
2                     A  
3                     A  
4                     A  
...                 ...  
2935                  Z  
2936                  Z  
2937                  Z  
2938                  Z  
2939                  Z  

[2940 rows x 57 columns]